Cubical/Structures/QuotientRing.agda

```diff
@@ -0,0 +1,248 @@
+    lemma x y a x-y∈I = eq/ (x + a) (y + a) (subst (λ u → u ∈ I) calculate x-y∈I)
+      where calculate : x - y ≡ (x + a) - (y + a)
+            calculate =
+                      x - y                 ≡⟨ translatedDifference a x y ⟩
+                      ((a + x) - (a + y))   ≡⟨ cong (λ u → u - (a + y)) (ring+-comm _ _) ⟩
+                      ((x + a) - (a + y))   ≡⟨ cong (λ u → (x + a) - u) (ring+-comm _ _) ⟩
+                      ((x + a) - (y + a))   ∎
+
+
+    homogenity' : (x y : R) → (x - y ∈ I)
+                  → translate x ≡ translate y
+    homogenity' x y x-y∈I i r = pointwise-equal r i
+      where
+        pointwise-equal : ∀ (u : R/I)
+                          → translate x u ≡ translate y u
+        pointwise-equal = elimProp (λ u → isSetR/I (translate x u) (translate y u))
+                                   (λ a → lemma x y a x-y∈I)
+
+    _+/I_ : R/I → R/I → R/I
+    x +/I y = (elim R/I→R/I-isSet translate homogenity' x) y
+      where
+        R/I→R/I-isSet = (λ r →  isOfHLevelΠ 2 (λ _ → squash/))
+
+    +/I-comm : (x y : R/I) → x +/I y ≡ y +/I x
+    +/I-comm = elimProp (λ x → isOfHLevelΠ 1 λ y → squash/ (x +/I y) (y +/I x))
+                                      (λ x' → elimProp (λ _ → squash/ _ _)
+                                                       λ y' → eq x' y')
+
+       where eq : (x y : R) → [ x ] +/I [ y ] ≡ [ y ] +/I [ x ]
+             eq x y i =  [ ring+-comm x y i ]
+
+    +/I-is-associative : (x y z : R/I) → x +/I (y +/I z) ≡ (x +/I y) +/I z
+    +/I-is-associative = elimProp
+                           (λ x → isOfHLevelΠ 1 λ y → isOfHLevelΠ 1 λ z → squash/ _ _)
+                           (λ x' → elimProp
+                                     (λ y → isOfHLevelΠ 1 λ z → squash/ _ _)
+                                     λ y' → elimProp (λ z → squash/ _ _)
+                                                     (λ z' → eq x' y' z'))
+
+      where eq : (x y z : R) → [ x ] +/I ([ y ] +/I [ z ]) ≡ ([ x ] +/I [ y ]) +/I [ z ]
+            eq x y z i =  [ ring+-assoc x y z i ]
+
+
+    0/I : R/I
+    0/I = [ ₀ ]
+
+    1/I : R/I
+    1/I = [ ₁ ]
+
+    -/I : R/I → R/I
+    -/I = elim (λ _ → squash/) (λ x' → [ - x' ]) eq
+      where
+        eq : (x y : R) → (x - y ∈ I) → [ - x ] ≡ [ - y ]
+        eq x y x-y∈I = eq/ (- x) (- y) (subst (λ u → u ∈ I) eq' (isIdeal.-closed I-isIdeal x-y∈I))
+          where
+            eq' = - (x + (- y))       ≡⟨ sym (-isDistributive _ _) ⟩
+                  (- x) - (- y)       ∎
+
+    +/I-rinv : (x : R/I) → x +/I (-/I x) ≡ 0/I
+    +/I-rinv = elimProp (λ x → squash/ _ _) eq
+      where
+        eq : (x : R) → [ x ] +/I (-/I [ x ]) ≡ 0/I
+        eq x i = [ ring+-rinv x i ]
+
+
+    +/I-rid : (x : R/I) → x +/I 0/I ≡ x
+    +/I-rid = elimProp (λ x → squash/ _ _) eq
+      where
+        eq : (x : R) → [ x ] +/I 0/I ≡ [ x ]
+        eq x i = [ ring+-rid x i ]
+
+    _·/I_ : R/I → R/I → R/I
+    _·/I_ =
+      elim (λ x → isOfHLevelΠ 2 (λ y → squash/))
```

Comment:
isSetPi2

Reply:
I use one "isSetPi" now, but I do not see how to use isSetPi2